### PR TITLE
Displaying Category image and Description

### LINKED ISSUE #3604: Display thumbnail for categories too
Thumbnails are super useful to distinguish two categories with similar names.

Many categories have images (for instance https://commons.wikimedia.org/wiki/Category:Rabbits see the upper right), probably taken from Wikidata.

We should display thumbnails, like is done for depictions.

### PR DESCRIPTION
**Pull request by Ayan-10**
**Description**

Fixes #3604 and #4527

**What changes did you make and why?**

Avail the images in for category for making it better distinguishable, getting thumbnails and description from API calls, and putting it in layout.

**Tests performed (required)**

Tested latest master on Pixel 3 API 30

**Screenshot**

### CODE REVIEW

**Comment by Ayan-10** [comment]
@nicolas-raoul Images are available now on display. Please review it once. And I decided to display the description in this PR. I am trying to find the API but I am not getting it. I will try again.

**Comment by Ayan-10** [comment]
@nicolas-raoul Should I use this API for getting the description?
https://commons.wikimedia.org/w/api.php?action=query&format=json&prop=description&titles=Category%3ARabbit%20(zodiac)

**Comment by nicolas-raoul** [comment]
Looks good indeed! :-)
If possible, make the loading asynchronous so that it does not slow down the rest of the information from showing.

**Comment by Ayan-10** [comment]
@nicolas-raoul I added description also. And I think I already made the call asynchronous and for me, It's not slowing down anything. Isn't it? 
If there is any problem with the PR, please let me know. I will try to fix that.

**Comment by nicolas-raoul** [comment]
It is very helpful and loads fast without slowing down anything, that's great!

Two things I have noticed:
- Many categories have the description "Wikimedia category", which is not super useful. In such cases maybe it would be possible to hop to the [category's main topic](https://www.wikidata.org/wiki/Q6527085)'s description? That can be left as a future enhancement, though.
- Descriptions and thumbnails could be cached a little longer? They seem to disappear as soon as I scroll, which does not happen for depictions. If you don't want to fix this now feel free to create a separate issue about this.

Thanks! It is making category selection much more reliable, and thus will make all uploaded pictures more findable by Wikipedia editors :-)

**Comment by Ayan-10** [comment]
@nicolas-raoul Yes, category's main topic's description will more useful. But I searched everywhere in MediaWiki APIs. I can't find any endpoint from where indirectly or directly I can get the category's main topics description. Can you give me some insight on how can I get category's main topic's description?

**Comment by Ayan-10** [comment]
2. I am telling you everything that I observed here. To get the thumbnail and description I need to do API calls. Now there 2 ways I can found to get thumbnail and description.

**1. BlockingGet**

**Pros:** The thumbnail and description will not appear and disappear with scrolling.
**Cons:** The UI will get very slow and buggy while fetching categories. Because BlockingGet will block the thread until it got the result which I don't think you wanted.

**2. Subscribe**

**Pros:** Will does not block any thread image and description will visible when it is will be available without any kind of lag in UI. So I followed this approach.
**Cons:** I think because of the above reason it is disappearing and appearing with scrolling. Because each item will try to fetch the thumbnail and description when the view will be attached to the data.

Now Depicts are not behaving like that because in depicts for thumbnail and description we are not doing other API calls. in depicts, every information will come with a single API call while loading and we are getting it from data class instance.

Still, I will try to do it in another way to make it more efficient (If possible).

Thanks for your appreciation.

**Comment by nicolas-raoul** [comment]
Understood for the difference with Depicts, thanks!

A "subscribe" strategy is good. I guess there is a way to cache its results momentarily, though?

**Comment by Ayan-10** [comment]
@nicolas-raoul I was thinking about getting the thumbnail and category along with searching the category. I mean like depicts, the API call will fetch category names along with thumbnail and description and display them when everything is available (how it's working for depicts). What's your thought on this idea? 

If you know any way to make it better, please let me know about that.

My exams are very close now so I am a little busy with them. I will start to work on it soon.

**Comment by nicolas-raoul** [comment]
It sounds like it would make the UI slower, unfortunately. For Depicts we
can get a lot of details with just one query.

How about maintaining in memory a
HashMap category→(thumbnail, description)
?

Of course, please prioritize your exams, that's the most important, good
luck! 👍

**Comment by Ayan-10** [comment]
@nicolas-raoul For categories also I think we can get all the details with thumbnail and description. 

See this: https://commons.wikimedia.org/w/api.php?action=query&format=json&formatversion=2&prop=categoryinfo%7Cdescription%7Cpageimages&piprop=thumbnail&pithumbsize=70&generator=allcategories&gacprefix=ra&gaclimit=25&gacoffset=0

I was talking about using this API call and do it like Depicts.

**Comment by nicolas-raoul** [comment]
Sorry for the delay!
The API call you found looks great, please feel free to proceed with it, thanks :-)

**Comment by Ayan-10** [comment]
@nicolas-raoul But If we proceed with this API we will get the thumbnail and description only in the production server. With the commons beta URL, we can't get thumbnail and description. But get can do this GET call with the prod URL in beta. Should I do that?

 I also have a PR regarding using prod URL for some beta GET calls. You can check #4551

**Comment by nicolas-raoul** [comment]
Production is the most important.

If the same does not work on the beta server, it is sad but should not prevent us from developing useful features like this one.

Afterwards we can continue the discussion at #4551

**Comment by Ayan-10** [comment]
@nicolas-raoul I made the required changes. You can review it now.

Thanks.

**Comment by Ayan-10** [comment]
Changed the logo. Is it ok now @nicolas-raoul?

**Comment by nicolas-raoul** [comment]
Thanks for changing the logo!

I am testing the PR now. It seems that the app does not see the categories I selected. Here I selected two categories, but strangely only one is visible on screen, and even more strangely I get the "No Categories Selected" popup:
adb
The uploaded file indeed does not have any of the two categories I selected.

**Comment by Ayan-10** [comment]
@nicolas-raoul I tested the PR on both beta and prod but Can't reproduce this kind of behavior. Can you please recheck it once?

**Comment by nicolas-raoul** [comment]
The problem seems to not happen every time, indeed.

**Comment by Ayan-10** [comment]
@nicolas-raoul Are you able to produce this issue on prod? I tested a lot of times and found out that sometimes in beta this issue is occurring. But in prod, It never happens for me. 

Actually, I am trying to find the cause of this unexpected issue, I was thinking that the cause might be the URL but I am not sure.

**Comment by nicolas-raoul** [comment]
Yes, that was in prod. Not sure if related, but I had a depiction selected, which resulted in a category being pre-selected.

**Comment by nicolas-raoul** [comment]
Sorry would you mind rebasing on master?

**Comment by nicolas-raoul** [comment]
Oops sorry you read m mind already! Thanks a lot, testing it :-)

**Comment by Ayan-10** [comment]
@nicolas-raoul I tried a lot to find the cause of this problem but can't find any change in my code which can occur this type of problem. I just added a description with `CategoryItem` everywhere in the codebase and show it as same as category name. Also, it's very hard to reproduce it so I can't test it properly, it occurred twice in 5 days for me. Can you check if this issue is happening with this PR or master too?

**Comment by nicolas-raoul** [comment]
I uploaded 3 pictures today and got the problem twice.
You are right, I will check whether it happens on master.

Le lun. 20 sept. 2021 à 14:13, Ayan Sarkar ***@***.***> a
écrit :

**Comment by Ayan-10** [comment]
Thanks, @nicolas-raoul please let me know about the master.  Don't know why it's very hard to reproduce for me. Can you upload some images for which this issue is occurring?

**Comment by Ayan-10** [comment]
@nicolas-raoul I changed something. Issues are not occurring for me but I am not sure as it's not occurring everytime. Can you check issues are still there or not?

**Comment by madhurgupta10** [comment]
@Ayan-10 Few of the tests are failing, can you take a look into that :)

Also, can you rebase your branch with the master?

```
> Task :app:kaptBetaDebugUnitTestKotlin
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/category/CategoriesModelTest.kt: (45, 25): Type mismatch: inferred type is List<String>! but List<CategoryItem>! was expected

e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/category/CategoriesModelTest.kt: (45, 37): Type mismatch: inferred type is List<String> but List<CategoryItem>! was expected
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/category/CategoriesModelTest.kt: (48, 65): The boolean literal does not conform to the expected type String
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/category/CategoriesModelTest.kt: (48, 70): No value passed for parameter 'thumbnail'
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/category/CategoriesModelTest.kt: (48, 70): No value passed for parameter 'isSelected'
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/category/CategoriesModelTest.kt: (68, 61): Type mismatch: inferred type is List<String> but List<CategoryItem> was expected
> Task :app:compileBetaDebugUnitTestKotlin FAILED
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/category/CategoriesModelTest.kt: (71, 25): Type mismatch: inferred type is List<String>! but List<CategoryItem>! was expected
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/category/CategoriesModelTest.kt: (71, 55): Type mismatch: inferred type is List<String> but List<CategoryItem>! was expected
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/category/CategoriesModelTest.kt: (79, 25): Type mismatch: inferred type is List<String>! but List<CategoryItem>! was expected
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/category/CategoriesModelTest.kt: (79, 37): Type mismatch: inferred type is List<String> but List<CategoryItem>! was expected
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/category/CategoriesModelTest.kt: (80, 63): Type mismatch: inferred type is List<String> but (Mutable)List<CategoryItem!> was expected
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/category/CategoryClientTest.kt: (37, 27): Type mismatch: inferred type is List<String> but List<CategoryItem>! was expected
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/category/CategoryClientTest.kt: (40, 27): Type mismatch: inferred type is List<String> but List<CategoryItem>! was expected
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/category/CategoryClientTest.kt: (62, 27): Type mismatch: inferred type is List<String> but List<CategoryItem>! was expected
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/category/CategoryClientTest.kt: (65, 27): Type mismatch: inferred type is List<String> but List<CategoryItem>! was expected
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/category/CategoryClientTest.kt: (87, 27): Type mismatch: inferred type is List<String> but List<CategoryItem>! was expected
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/category/CategoryClientTest.kt: (107, 27): Type mismatch: inferred type is List<String> but List<CategoryItem>! was expected
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/upload/CategoriesPresenterTest.kt: (70, 92): The boolean literal does not conform to the expected type String
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/upload/CategoriesPresenterTest.kt: (76, 68): The boolean literal does not conform to the expected type String
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/upload/GpsCategoryModelTest.kt: (23, 64): Type mismatch: inferred type is List<String> but List<CategoryItem> was expected
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/upload/GpsCategoryModelTest.kt: (24, 40): Type mismatch: inferred type is List<String> but List<CategoryItem>! was expected
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/utils/StringSortingUtilsTest.kt: (15, 26): Type mismatch: inferred type is String! but CategoryItem! was expected
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/utils/StringSortingUtilsTest.kt: (41, 26): Type mismatch: inferred type is String! but CategoryItem! was expected
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/utils/StringSortingUtilsTest.kt: (63, 26): Type mismatch: inferred type is String! but CategoryItem! was expected
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/utils/StringSortingUtilsTest.kt: (93, 26): Type mismatch: inferred type is String! but CategoryItem! was expected
e: /home/runner/work/apps-android-commons/apps-android-commons/app/src/test/kotlin/fr/free/nrw/commons/utils/StringSortingUtilsTest.kt: (123, 26): Type mismatch: inferred type is String! but CategoryItem! was expected
```

**Comment by nicolas-raoul** [comment]
Despite the [Paris](https://commons.wikimedia.org/wiki/Category:Paris) category having both an image and a description, I do not see them on the app with the current code of this branch, is it intended? Thanks! :-)

Screen Shot 2021-12-22 at 18 13 17

adb

**Comment by Ayan-10** [comment]
@nicolas-raoul I can't reproduce the issue that you are facing. As you can see in the screenshot Category Paris is appearing with an image and description. 

But I can see the UI is not working properly now. Don't know how? It was fine when I checked it last time. I will fix the UI.

**Comment by nicolas-raoul** [comment]
Maybe I had a network problem or something.
Understood, I will test again after your fix.
Thanks!

**Comment by Ayan-10** [comment]
@nicolas-raoul I fixed the UI issue. Also, another issue that you are facing is actually occurring. When you will search Paris in the search box it will not occur, as you can see in my screenshot the image of Paris is showing as expected. But when the Paris is coming from a selected depiction (without searching Paris), it will not show the thumbnail and description because when we are getting categories from depiction the thumbnail and description is not available in the server. This is the same problem that I mentioned in  #4539 .

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
Issue:
1. Upload an image
2. At the "depictions" step, select no depictions
3. At the "categories" step, type "paris" and select the "Paris" category
4. Press your keyboard's backspace 5 times to remove the search string
5. Problem 1: No selected categories are visible
6. Type "paris" again
7. Problem 2: The category we selected at step 3 is not selected anymore
8. Press "Next"
9. Problem 3: "No categories" selected appears, and indeed after uploading the picture does not have the "Paris" category.

Not sure whether this is 3 different problems or different facets of the same one.
These problems do not happen in master.

Thanks! By the way, the images look very useful to choose the right category :-)

**Comment by Ayan-10** [comment]
@nicolas-raoul The problem is it's not occurring every time. I tried with the exact same steps but it's occurred only once. Also, I didn't change any logic in selecting categories, don't know why it's happening. And as it's very hard to reproduce and very randomly occurring, I am not able to test to find the cause of the issue. Still, I fixed something to fix the issue. Please review it once. And also If you observe any specific event for which the issue is occurring please let me know.

Thanks.

**Review by neslihanturan — CHANGES_REQUESTED** [review]
Hi @Ayan-10 , unfortunately I can also reproduce the issue that @nicolas-raoul  mentioned :(

**Comment by Ayan-10** [comment]
@nicolas-raoul @neslihanturan I fixed the issue (selected categories disappearing). Can you please check it once?

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
Using this branch, I don't see any depiction suggestion, even when typing something in the search box.

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
Without any depiction selected, on the "Categories" step of the upload wizard, before even typing anything, the spinner at the right of the category search bar spins forever. Typing anything in the search bar does not change anything.

**Comment by Ayan-10** [comment]
@nicolas-raoul I checked those issues. I was on another PR. I installed this PR (without uninstalling the app with another PR). These issues occurred for me. Then I uninstalled the app and then installed this PR, after that these issues are not occurring anymore in this PR. I rechecked by doing the same process and the same thing happened (Issues are occurring if I install the app without uninstalling the previous version).

Can you please check it once after uninstalling the current version and then installing this PR?

Thanks.

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
Indeed, thanks for the tip!

It works great :-)

I just noticed one small bug so far, where the categories I had selected got deselected. will try to find a way to reproduce the issue, please try to reproduce it too, thanks!

File: gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -1,6 +1,6 @@
-#Thu Dec 26 14:27:31 IST 2019
+#Thu Jul 29 23:46:38 IST 2021
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
 zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https://services.gradle.org/distributions/gradle-6.9-all.zip
+distributionUrl=https://services.gradle.org/distributions/gradle-6.9-all.zip
```
Comment (nicolas-raoul) [inline gradle/wrapper/gradle-wrapper.properties:6]: Please revert changes in this file, thanks!

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
Please post a screenshot showing how the categories selection screen look like on Android API level 30, thanks!

**Comment by nicolas-raoul** [comment]
Thanks Ayan for finding this out!
Would you mind creating a new issue about it?

**Comment by Ayan-10** [comment]
@nicolas-raoul The issue has been created.

**Review by neslihanturan — APPROVED** [review]
I am not able to test this PR on API 19 (because of the issue already reported) but it works for higher APIs for me. We can either merge now or wait until API 19 issue is solved and we can test it too.

File: app/src/main/java/fr/free/nrw/commons/bookmarks/items/BookmarkItemsDao.java [hunk 2]
```diff
@@ -60,6 +64,7 @@ public BookmarkItemsDao(
      * @param depictedItem : Bookmark object
      * @return boolean : is bookmark now favorite ?
      */
+    @RequiresApi(api = VERSION_CODES.N)
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/bookmarks/items/BookmarkItemsDao.java:67]: Could you please add a javadoc comment explaining why?

File: app/src/main/java/fr/free/nrw/commons/bookmarks/items/BookmarkItemsDao.java [hunk 3]
```diff
@@ -74,6 +79,7 @@ public boolean updateBookmarkItem(final DepictedItem depictedItem) {
      * Add a Bookmark to database
      * @param depictedItem : Bookmark to add
      */
+    @RequiresApi(api = VERSION_CODES.N)
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/bookmarks/items/BookmarkItemsDao.java:82]: Could you please leave a comment explaining why?

File: app/src/main/java/fr/free/nrw/commons/bookmarks/items/BookmarkItemsDao.java [hunk 4]
```diff
@@ -187,13 +212,19 @@ private String ArrayToString(final List<String> list) {
      * @param depictedItem depicted item
      * @return ContentValues
      */
+    @RequiresApi(api = VERSION_CODES.N)
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/bookmarks/items/BookmarkItemsDao.java:215]: Could you please leave a comment explaining why?

File: app/src/main/java/fr/free/nrw/commons/category/CategoryEditSearchRecyclerViewAdapter.java [hunk 5]
```diff
@@ -91,11 +94,15 @@ public void addToNewCategories(String addedCategory) {
     public Filter getFilter() {
         return new Filter() {
 
+            @RequiresApi(api = VERSION_CODES.N)
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/category/CategoryEditSearchRecyclerViewAdapter.java:97]: Could you please leave a comment explaining why?

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
Could you please add a screenshot showing how it looks on API level 23 (version code M)? Thanks!

**Comment by Ayan-10** [comment]
@nicolas-raoul I was using stream() function for java classes to extract and map objects and as stream() is only supported till API level N, I need that annotation. Now I removed all steam() functions and did it in another way so API level doesn't matter anymore.

Thanks

**Comment by nicolas-raoul** [comment]
I also love streams!
Unfortunately we have to support old versions of Android, as many of our most important users do not have the financial means to use anything but old phones.
Thanks for the modification!

**Review by nicolas-raoul — COMMENTED** [review]
It is generally working great!

Would you know why the category [TH Liner](https://commons.wikimedia.org/wiki/Category:TH_Liner) does not show an icon nor description?

adb

**Review by nicolas-raoul — APPROVED** [review]

**Comment by Ayan-10** [comment]
@nicolas-raoul I hope the TH Liner is coming from selected depictions that you selected in the previous screen. That is why it's not showing any other information other than the name. I explained the reason in [this comment](https://github.com/commons-app/apps-android-commons/pull/4531#issuecomment-1000094870).

Other than that If you search TH Liner in the search box it will show the description and thumbnail. As you can see in the screenshot.
Screenshot_1642657636

**Comment by nicolas-raoul** [comment]
Answering my TH Liner question above:

It is probably because of using the Mediawiki API's [pageimages](https://www.mediawiki.org/wiki/Extension:PageImages) method, which does not get images from [{{Wikidata Infobox}}](https://commons.wikimedia.org/w/index.php?title=Category:TH_Liner&action=edit).

An improvement would be getting the P18 from [the Wikidata item of the category](https://www.wikidata.org/wiki/Q85867756) if `pageimages` returns nothing. Do you want to try? If not, I can just merge this for now, and create an enhancement issue. Like you want :-)

**Comment by nicolas-raoul** [comment]
Ah sorry I just saw your reply! Oh, looks like my intuition was wrong :-)

**Comment by nicolas-raoul** [comment]
Thank you for linking to https://github.com/commons-app/apps-android-commons/pull/4531#issuecomment-1000094870 and sorry for forgetting that!
In such cases, image/description could be retrieved from Wikidata. Do you want to try, or should I merge this as is, and create a separate issue?

**Comment by nicolas-raoul** [comment]
And to answer your question, yes, TH Liner came because I had selected the TH Liner depiction :-)

**Comment by Ayan-10** [comment]
I would like to work on that as a separate issue. 

Thanks.

**Comment by nicolas-raoul** [comment]
Understood. Would you mind creating the issue?
Thanks for this very useful PR!

**Comment by Ayan-10** [comment]
Sure, I will create the issue.

**Pull request merged**